File: baml_language/sdk_tests/crates/rust/type_shapes/customizable/roundtrip_tests/test_optional.rs
//! Roundtrip coverage for `baml_sdk::optional` — optional Ty variants.

use baml_sdk::optional::{
    IntOrString, OptionalContainer, Resume, round_trip_optional_container, round_trip_optional_int,
    round_trip_optional_resume, round_trip_optional_union, round_trip_resume,
};

#[test]
fn test_optional_round_trip_optional_int() {
    assert_eq!(round_trip_optional_int(Some(5)).unwrap(), Some(5));
    assert_eq!(round_trip_optional_int(None).unwrap(), None);
}

#[test]
fn test_optional_round_trip_optional_resume() {
    let r = Resume {
        name: "ada".to_string(),
    };
    assert_eq!(
        round_trip_optional_resume(Some(r.clone())).unwrap(),
        Some(r)
    );
    assert_eq!(round_trip_optional_resume(None).unwrap(), None);
}

#[test]
fn test_optional_round_trip_optional_union() {
    assert_eq!(
        round_trip_optional_union(Some(IntOrString::Int(3))).unwrap(),
        Some(IntOrString::Int(3))
    );
    assert_eq!(
        round_trip_optional_union(Some(IntOrString::String("s".to_string()))).unwrap(),
        Some(IntOrString::String("s".to_string()))
    );
    assert_eq!(round_trip_optional_union(None).unwrap(), None);
}

#[test]
fn test_optional_round_trip_resume() {
    let r = Resume {
        name: "grace".to_string(),
    };
    assert_eq!(round_trip_resume(r.clone()).unwrap(), r);
}

#[test]
fn test_optional_round_trip_optional_container() {
    let c = OptionalContainer {
        optional_int: None,
        optional_class: Some(Resume {
            name: "x".to_string(),
        }),
        optional_union: Some(IntOrString::String("y".to_string())),
    };
    assert_eq!(round_trip_optional_container(c.clone()).unwrap(), c);
}

File: baml_language/sdk_tests/crates/python_pydantic2/type_shapes/customizable/roundtrip_tests/test_optional.py
"""Roundtrip coverage for `baml_sdk.optional` — optional Ty variants."""

import baml_sdk  # noqa: F401  — initializes the BAML runtime
from baml_sdk.optional import (
    Resume,
    OptionalContainer,
    round_trip_optional_int,
    round_trip_optional_resume,
    round_trip_optional_union,
    round_trip_resume,
    round_trip_optional_container,
)


def test_optional_round_trip_optional_int():
    assert round_trip_optional_int(x=5) == 5
    assert round_trip_optional_int(x=None) is None


def test_optional_round_trip_optional_resume():
    r = Resume(name="ada")
    assert round_trip_optional_resume(r=r) == r
    assert round_trip_optional_resume(r=None) is None


def test_optional_round_trip_optional_union():
    assert round_trip_optional_union(u=3) == 3
    assert round_trip_optional_union(u="s") == "s"
    assert round_trip_optional_union(u=None) is None


def test_optional_round_trip_resume():
    r = Resume(name="grace")
    assert round_trip_resume(r=r) == r


def test_optional_round_trip_optional_container():
    c = OptionalContainer(
        optional_int=None,
        optional_class=Resume(name="x"),
        optional_union="y",
    )
    assert round_trip_optional_container(c=c) == c
